packages/admin/cms-admin/src/blocks/InternalLinkBlock.tsx

```diff
@@ -62,19 +75,131 @@ export const InternalLinkBlock: BlockInterface<InternalLinkBlockData, State, Int
+                                            newState.targetPageAnchor === "none" || newState.targetPageAnchor === ""
+                                                ? undefined
+                                                : newState.targetPageAnchor,
+                                    };
+                                }
+                            });
+                        }}
+                        initialValues={{
+                            targetPage: state.targetPage,
+                            targetPageAnchor: anchorsLoading ? "" : state.targetPageAnchor ?? "none",
+                        }}
+                    >
+                        <Field name="targetPage" component={FinalFormPageTreeSelect} fullWidth fieldContainerProps={{ fieldMargin: "never" }} />
+                        <Divider />
+                        <Box padding={3}>
+                            <Field
+                                name="targetPageAnchor"
+                                label={<FormattedMessage id="comet.blocks.internalLink.anchor.label" defaultMessage="Anchor" />}
+                                fullWidth
+                                disabled={state.targetPage == null || anchorsLoading}
+                            >
+                                {(props) => (
+                                    <FinalFormSelect {...props} disabled={state.targetPage == null || anchorsLoading}>
+                                        <MenuItem value="none">
+                                            <FormattedMessage id="comet.blocks.internalLink.anchor.none" defaultMessage="None" />
+                                        </MenuItem>
+                                        {anchors?.map((anchor) => (
+                                            <MenuItem key={anchor} value={anchor}>
+                                                {anchor}
+                                            </MenuItem>
+                                        ))}
+                                    </FinalFormSelect>
+                                )}
+                            </Field>
+                        </Box>
+                    </BlocksFinalForm>
+                </AdminComponentPaper>
             </SelectPreviewComponent>
         );
     },
 
-    previewContent: (state) => (state.targetPage?.name ? [{ type: "text", content: state.targetPage.name }] : []),
+    previewContent: (state) => {
+        if (state.targetPage == null) {
+            return [];
+        }
+
+        return [
+            {
+                type: "text",
+                content: state.targetPageAnchor === undefined ? state.targetPage.name : `${state.targetPage.name}#${state.targetPageAnchor}`,
+            },
+        ];
+    },
 };
+
+function useTargetPageAnchors(targetPage: { id: string; documentType: string } | null | undefined): string[] | undefined {
+    const [anchors, setAnchors] = React.useState<string[]>();
```

Comment:
do you really need to duplicate state?

Reply:
What do you mean by duplicate state? Anchors are only defined here

Reply:
I mean, Apollo Cache is not the SSOT here while I think it could be.

Reply:
How would I do that? I can't use `useQuery` because the target page or the document type may be undefined...

Reply:
Ok, I see the problem. Maybe in combination with skip:true (and somehow silencing type errors)

Reply:
I've tried that, but useQuery fail when I provide undefined for `query`. Might add a noop query, then it would work. Is this a good idea?

Reply:
> I've tried that, but useQuery fail when I provide undefined for `query`. Might add a noop query, then it would work. Is this a good idea?

I've tried this in 814050933ea9335ab94f643c0ae3c2ddf7563a8b, and it works: What do you think?

Reply:
interesting hack :)

while it is hackish, I prefer it over state duplication :+1: